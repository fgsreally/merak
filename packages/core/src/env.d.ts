import { Merak } from './merak';
declare global {
    interface Window {
        dImport: (url:string)=>void
        $Merak:any
        $MerakMap:any
        $MerakBus:any
        __merak_url__?:string
        __merak_app__?:Merak
    }
}


export { }