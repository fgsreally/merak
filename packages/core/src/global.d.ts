
import { Merak } from './merak';
declare global {
    interface Window {
       
      
        MERAK_DEBUG:string
    }
}


export { }