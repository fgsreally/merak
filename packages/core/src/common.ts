export const MERAK_DATA_ID = 'data-merak-id'
export const MERAK_KEEP_ALIVE = 'keep-alive'
export const MERAK_EVENT_PREFIX = 'merak_'
// export const MERAK_EVENT_EXEC_SCRIPT = `${MERAK_EVENT_PREFIX}exec_script`
export const MERAK_EVENT_RELUNCH = `${MERAK_EVENT_PREFIX}relunch`
export const MERAK_EVENT_MOUNT = `${MERAK_EVENT_PREFIX}mount`
export const MERAK_EVENT_HIDDEN = `${MERAK_EVENT_PREFIX}hidden`
export const MERAK_EVENT_DESTROY = `${MERAK_EVENT_PREFIX}destroy`

export const MERAK_SHADE_STYLE = 'position: fixed; z-index: 111111111; visibility: hidden; inset: 0px; backface-visibility: hidden;'
