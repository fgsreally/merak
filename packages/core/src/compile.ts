import MagicString from 'magic-string'

import type { ImportFile, ImportScript, InlineScript } from 'merak-compile'
import { desctructGlobal, resolveUrl } from './utils'

export function compileGlobals(s: MagicString, globals: string[], insertPos: number, fakeGlobalVar: string) {
  s.appendRight(insertPos || 0, `\nconst {${desctructGlobal(globals)}}=${fakeGlobalVar};\n`)
  return s
}

export function compileStaticImports(s: MagicString, staticImport: ImportFile[], baseURL: string, urlMap: Map<string, string>) {
  for (const importInfo of staticImport) {
    const url = resolveUrl(importInfo.filePath, baseURL)
    const virtualUrl = urlMap.get(url)

    if (!virtualUrl)
      throw new Error(JSON.stringify({ stage: 'static-import', filePath: url, info: 'can\'t get file\'s virtual url ' }))
    const [start, end] = importInfo.loc

    s.overwrite(start, end, `"${virtualUrl}"`)
  }

  return s
}

export function compileDynamicImports(s: MagicString, dynamicImport: { [filePath in string]: { loc: [number, number] } }, projectID: string) {
  for (const i in dynamicImport) {
    const [start, end] = dynamicImport[i].loc
    s.overwrite(start, end, `$MerakImport('${i}','${projectID}')`)
  }

  return s
}

export function compileHTMLJS(s: MagicString, url: string, files: (ImportScript | InlineScript)[], globals: string[], fakeGlobalVar: string) {
  for (const file of files) {
    if (file._t !== 'outline') {
      const [tagStart, tagEnd] = (file as InlineScript)._b

      const body = new MagicString(s.slice(tagStart, tagEnd))
      for (const [start, end] of file._l) {
        const importer = body.slice(start + 1, end - 1)
        body.overwrite(start, end, `"${resolveUrl(importer, url)}"`)
      }
      if (file._t === 'esm')
        file._a.innerHTML = `\nconst {${desctructGlobal(globals)}}=${fakeGlobalVar};\n${body.toString()}`
      if (file._t === 'iife')
        file._a.innerHTML = `(()=>{const {${desctructGlobal(globals)}}=${fakeGlobalVar};${body.toString()}})()`
    }
    else {
      file._a.src = file._f
    }

    const [start, end] = file._tl
    s.overwrite(start, end, '')
  }
}

export function compileInlineJS(s: MagicString, scripts: { loc: [number, number]; type: 'esm' | 'iife' }[], globals: string[], fakeGlobalVar: string) {
  for (const script of scripts) {
    if (script.type === 'esm')
      s.appendRight(script.loc[0], `\nconst {${desctructGlobal(globals)}}=${fakeGlobalVar};\n`)

    if (script.type === 'iife') {
      s.appendRight(script.loc[0], `(()=>{const {${desctructGlobal(globals)}}=${fakeGlobalVar};\n`)
      s.appendLeft(script.loc[1], '})()')
    }
  }
  return s
}

export function compileCSSLink(s: MagicString, links: { [filePath in string]: { _l: [number, number] } }, baseURL: string) {
  for (const i in links) {
    const [start, end] = links[i]._l
    s.overwrite(start, end, `"${resolveUrl(i, baseURL)}"`)
  }
}
