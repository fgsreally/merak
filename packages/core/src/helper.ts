import type { Merak } from './merak'

export const MerakMap: Map<string, Merak> = window.$MerakMap || new Map()
window.$MerakMap = MerakMap

export function getInstance(id: string) {
  return MerakMap.get(id)
}

export function getHost(id: string) {
  return getInstance(id)?.shadowRoot.host
}

export function getUrl(id: string) {
  return (getInstance(id)?.proxyMap.location as Location)?.href
}

let MERAK_BODY_STYLE = 'position:relative'
export function setBodyStyle(style: string) {
  MERAK_BODY_STYLE = style
}

export function getBodyStyle() {
  return MERAK_BODY_STYLE
}
