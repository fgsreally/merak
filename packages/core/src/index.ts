export * from './common'
export * from './helper'
export * from './merak'
export * from './shadow'
export * from './types'
export * from './utils'
export * from './proxy'
export * from './perf'
export * from './iframe'
