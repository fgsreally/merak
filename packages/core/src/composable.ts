import { MERAK_DATA_FAKEGLOBALVAR, MERAK_DATA_ID } from './common'
import { Merak } from './merak'
import type { ProxyGlobals } from './types'

export function execScript(idAndName: string, url: string, proxy: ProxyGlobals) {
  if (!getInstance(idAndName)) {
  // eslint-disable-next-line @typescript-eslint/indent, no-new
  new Merak(idAndName, url, { proxy })
    const el = document.createElement('merak-block')
    el.setAttribute(MERAK_DATA_ID, idAndName)
    el.setAttribute(MERAK_DATA_FAKEGLOBALVAR, idAndName)
    document.body.appendChild(el)
  }

  return import(url)
}

export const MerakMap: Map<string, Merak> = window.$MerakMap || new Map()
window.$MerakMap = MerakMap

// export const bus = window.$MerakBus || mitt()
// window.$MerakBus = bus

export function getInstance(id: string) {
  return MerakMap.get(id)
}

export function getHost(id: string) {
  return getInstance(id)?.shadowRoot.host
}

export function getUrl(id: string) {
  return (getInstance(id)?.proxyMap.location as Location)?.href
}

let MERAK_BODY_STYLE = 'position:relative'
export function setBodyStyle(style: string) {
  MERAK_BODY_STYLE = style
}

export function getBodyStyle() {
  return MERAK_BODY_STYLE
}
