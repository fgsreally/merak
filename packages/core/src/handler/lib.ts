import { MERAK_DATA_ID, MERAK_DATA_VARNAME } from '../common'
import { getInstance } from '../composable'
import { Merak } from '../merak'
import type { CustomProxyHandler } from '../types'
import { flow, noop } from '../utils'
export const libHandler: CustomProxyHandler = (arg: any) => {
  delete arg.history
  delete arg.location
  return arg
}

export function execScript(idAndName: string, url: string, customHandler: CustomProxyHandler = noop) {
  if (!getInstance(idAndName)) {
  // eslint-disable-next-line @typescript-eslint/indent, no-new
  new Merak(idAndName, url, { customHandler: flow(libHandler, customHandler) })
    const el = document.createElement('merak-block')
    el.setAttribute(MERAK_DATA_ID, idAndName)
    el.setAttribute(MERAK_DATA_VARNAME, idAndName)
    document.body.appendChild(el)
  }

  return import(url)
}
