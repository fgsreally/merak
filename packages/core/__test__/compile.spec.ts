import fs from 'fs'
import { resolve } from 'path'
import { describe, expect, it } from 'vitest'
import { analyseHTML } from 'merak-compile'
import { compileHTML } from '../src/compile'
describe('analyse file', () => {
  it('compileInlineJS', async () => {
    const content = await fs.promises.readFile(resolve(__dirname, './fixtures/index.html'), 'utf-8')
    const ret = analyseHTML(content)
    expect(compileHTML(content, 'http://localhost:5002', ret._l)).toMatchSnapshot()
  })

  // it('compile esm js', async () => {
  //   const filePath = resolve(__dirname, './fixtures/esm.js')
  //   const { code } = injectGlobalToESM(await fs.promises.readFile(filePath, 'utf-8'), '$test', ['document', 'window'])
  //   expect(code).toMatchSnapshot()
  // })
})
