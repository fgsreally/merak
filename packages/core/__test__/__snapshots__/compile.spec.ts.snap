// Vitest Snapshot v1

exports[`analyse file > compileInlineJS 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <link rel=\\"icon\\" href=\\"http://localhost:5002/favicon.ico\\" />
    <link rel=\\"stylesheet\\" href=\\"http://localhost:5002/assets/index-28fadf32.css\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
  </head>

  <body>
    <script type=\\"module\\" src=\\"http://localhost:5002/src/main.js\\"></script>
    <script type=\\"module\\">
      import y from 'y'
import xx from 'http://localhost:5002/xx'
document.write(xx)
console.log(y)
    </script>
    <script>
      window.addEventListener('click', async () => {
        await import('http://localhost:5002/esm.js')
})
    </script>
  </body>
</html>
"
`;
